Base/QTGUI/qSlicerExportNodeDialog.cxx

```diff
@@ -0,0 +1,651 @@
+
+  QWidget* optionsContainerChild = this->OptionsContainer->findChild<QWidget*>(QString(), Qt::FindDirectChildrenOnly);
+  if (!optionsContainerChild)
+    return nullptr;
+
+  qSlicerFileWriterOptionsWidget* options = dynamic_cast<qSlicerFileWriterOptionsWidget*>(optionsContainerChild);
+  Q_ASSERT(options != nullptr);
+  return options;
+}
+
+//-----------------------------------------------------------------------------
+void qSlicerExportNodeDialogPrivate::formatChangedSlot()
+{
+  this->formatChanged(true);
+}
+
+//-----------------------------------------------------------------------------
+void qSlicerExportNodeDialogPrivate::formatChanged(bool updateFilename)
+{
+  // In case the combobox was editable (hack to display custom text), we now
+  // don't need this property anymore.
+  this->ExportFormatComboBox->setEditable(false);
+
+  if (updateFilename)
+    this->FilenameLineEdit->setText(this->recommendedFilename());
+
+  this->updateOptionsWidget();
+}
+
+//-----------------------------------------------------------------------------
+QString qSlicerExportNodeDialogPrivate::recommendedFilename() const
+{
+  // Get the extension associated with the currently selected format
+  QString extension = QString::fromStdString(vtkDataFileFormatHelper::GetFileExtensionFromFormatString(
+    this->formatText().toUtf8()));
+  if (extension == "*")
+    extension = QString();
+
+  return forceFileNameExtension(this->FilenameLineEdit->text(), extension, this->MRMLNode);
+}
+
+//-----------------------------------------------------------------------------
+void qSlicerExportNodeDialogPrivate::updateOptionsWidget()
+{
+  // If there is already an options widget corresponding to the format currently being used,
+  // then just use that. Otherwise we go ahead and create a new options widget.
+  auto iterator = this->formatToOptionsWidget.find(this->formatText());
+  if (iterator != this->formatToOptionsWidget.end()) {
+    this->setOptionsWidget(*iterator);
+    return;
+  }
+
+
+  // Create new options widget for the present node type
+  qSlicerCoreIOManager* coreIOManager =
+    qSlicerCoreApplication::application()->coreIOManager();
+  qSlicerIOOptions* options =
+    coreIOManager->fileWriterOptions(this->MRMLNode, this->formatText());
+  qSlicerFileWriterOptionsWidget* optionsWidget =
+    dynamic_cast<qSlicerFileWriterOptionsWidget*>(options);
+
+  // If the dynamic cast succeeds, we will pass ownership of the options widget
+  // to the export dialog.
+  // Otherwise, we will delete the options widget right now, since we cannot use
+  // options that are not widgets.
+
+  if (optionsWidget)
+  {
+    // The optionsWidget can use the filename to initialize some options.
+    optionsWidget->setFileName(this->file().absoluteFilePath());
+    optionsWidget->setObject(this->MRMLNode);
+
+    this->setOptionsWidget(optionsWidget); // we will pass ownership of optionsWidget to Qt
```

Comment:
Yes that's what I mean. I mean to say that I'm relying on the memory management built into QWidget hierarchies. The widget returned by `qSlicerCoreIOManager::fileWriterOptions` in

```cpp
qSlicerIOOptions* options =
    coreIOManager->fileWriterOptions(this->MRMLNode, this->formatText());
```

has no owner managing its memory.

Maybe it would be clearer if I made the comment say "we will pass ownership of optionsWidget to the qSlicerExportNodeDialog widget" ?

Reply:
> Maybe it would be clearer if I made the comment say "we will pass ownership of optionsWidget to the qSlicerExportNodeDialog widget" ?

Yes